File: app/src/main/java/li/klass/fhem/fragments/ConversionFragment.java
/*
 * AndFHEM - Open Source Android application to control a FHEM home automation
 * server.
 *
 * Copyright (c) 2011, Matthias Klass or third-party contributors as
 * indicated by the @author tags or express copyright attribution
 * statements applied by the authors.  All third-party contributions are
 * distributed under license by Red Hat Inc.
 *
 * This copyrighted material is made available to anyone wishing to use, modify,
 * copy, or redistribute it subject to the terms and conditions of the GNU GENERAL PUBLIC LICENSE, as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU GENERAL PUBLIC LICENSE
 * for more details.
 *
 * You should have received a copy of the GNU GENERAL PUBLIC LICENSE
 * along with this distribution; if not, write to:
 *   Free Software Foundation, Inc.
 *   51 Franklin Street, Fifth Floor
 *   Boston, MA  02110-1301  USA
 */

package li.klass.fhem.fragments;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;

import li.klass.fhem.R;
import li.klass.fhem.fragments.core.BaseFragment;
import li.klass.fhem.util.NumberSystemUtil;

public class ConversionFragment extends BaseFragment {

    private transient EditText inputField;
    private transient TextView resultField;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        super.onCreateView(inflater, container, savedInstanceState);

        View view = inflater.inflate(R.layout.conversion, container, false);
        inputField = (EditText) view.findViewById(R.id.input);
        Button hexToQuatButton = (Button) view.findViewById(R.id.hexToQuat);
        Button quatToHexButton = (Button) view.findViewById(R.id.quatToHex);
        resultField = (TextView) view.findViewById(R.id.result);

        hexToQuatButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                try {
                    String input = getInputText();
                    String result = NumberSystemUtil.hexToQuaternary(input, 0);
                    setResult(result);
                } catch (IllegalArgumentException e) {
                    setResult(getActivity().getString(R.string.error));
                }
            }
        });
        quatToHexButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                try {
                    String input = getInputText();
                    String result = NumberSystemUtil.quaternaryToHex(input);
                    setResult(result);
                } catch (IllegalArgumentException e) {
                    setResult(getActivity().getString(R.string.error));
                }
            }
        });
        return view;
    }

    private String getInputText() {
        return inputField.getText().toString();
    }

    private void setResult(String result) {
        resultField.setText(result);
    }

    @Override
    public void update(boolean doUpdate) {
    }
}

File: app/src/main/java/li/klass/fhem/conversion/ui/ConversionFragment.kt
/*
 * AndFHEM - Open Source Android application to control a FHEM home automation
 * server.
 *
 * Copyright (c) 2011, Matthias Klass or third-party contributors as
 * indicated by the @author tags or express copyright attribution
 * statements applied by the authors.  All third-party contributions are
 * distributed under license by Red Hat Inc.
 *
 * This copyrighted material is made available to anyone wishing to use, modify,
 * copy, or redistribute it subject to the terms and conditions of the GNU GENERAL PUBLIC LICENSE, as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU GENERAL PUBLIC LICENSE
 * for more details.
 *
 * You should have received a copy of the GNU GENERAL PUBLIC LICENSE
 * along with this distribution; if not, write to:
 *   Free Software Foundation, Inc.
 *   51 Franklin Street, Fifth Floor
 *   Boston, MA  02110-1301  USA
 */

package li.klass.fhem.conversion.ui

import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import li.klass.fhem.R
import li.klass.fhem.databinding.ConversionBinding
import li.klass.fhem.fragments.core.BaseFragment
import li.klass.fhem.util.NumberSystemUtil
import javax.inject.Inject

class ConversionFragment @Inject constructor(): BaseFragment() {

    private lateinit var binding: ConversionBinding
    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        super.onCreateView(inflater, container, savedInstanceState)

        binding = ConversionBinding.inflate(inflater, container, false)

        val inputField = binding.input
        val hexToQuatButton = binding.hexToQuat
        val quadToHexButton = binding.quatToHex

        hexToQuatButton.setOnClickListener {
            try {
                val input = inputField.text.toString()
                val result = NumberSystemUtil.hexToQuaternary(input, 0)
                setResult(result)
            } catch (e: IllegalArgumentException) {
                setResult(requireActivity().getString(R.string.error))
            }
        }
        quadToHexButton.setOnClickListener {
            try {
                val input = inputField.text.toString()
                val result = NumberSystemUtil.quaternaryToHex(input)
                setResult(result)
            } catch (e: IllegalArgumentException) {
                setResult(requireActivity().getString(R.string.error))
            }
        }
        return binding.root
    }

    private fun setResult(result: String) {
        binding.result.text = result
    }

    override fun mayPullToRefresh(): Boolean = false

    override fun getTitle(context: Context) = context.getString(R.string.conversion)

    override suspend fun update(refresh: Boolean) {}
}
